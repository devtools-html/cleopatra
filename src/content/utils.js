// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
