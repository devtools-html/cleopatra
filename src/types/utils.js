/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

/*
 * A not-void type
 * https://flow.org/try/#0C4TwDgpgBAcg9sAanAlgEygXigKClAHygDsBXAWwCMIAnPQqAZ2BpWIHN6jK44AbCAENiXEqT59RAQRo1BIADzCQAPlEB5SgCsIAY2ABuHDgBmpYvpRxiUMxYUAVAFywEydCoAUAN0F9SEC4OAJRBUADe9DQQwKQ0Nr7+EEYAvsZ2up4AjMFGGZ4ARMysHAW5puaZ4ba8LgDklII0dVAp5flkEu2VngDaWQA0UABMALrdFp7maBAmbBBoE5nlQA
 */
export type NotVoidOrNull = number | string | boolean | Array<any> | Object;

export type ExtractReturnType = <V>((...args: any[]) => V) => V;

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
// eslint-disable-next-line no-unused-vars
export type IndexedArray<_IndexType, Value> = Array<Value>;

/**
 * This is a utility type that extracts the return type of a function.
 */
export type $ReturnType<Fn> = $Call<ExtractReturnType, Fn>;
