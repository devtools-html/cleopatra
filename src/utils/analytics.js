/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Document Google Analytics API that is used in the project. These definitions
 * can be updated as more features are used.
 *
 * https://developers.google.com/analytics/devguides/collection/analyticsjs/pages
 */
type GAEvent = {|
  hitType: 'event',
  // Specifies the event category. Must not be empty
  eventCategory: string,
  eventAction: string,
  eventLabel?: string,
  eventValue?: number,
|};

type GAPageView = {|
  hitType: 'pageview',
  page: string,
|};

type GATiming = {|
  hitType: 'timing',
  timingCategory: string,
  timingVar: string,
  timingValue: number,
  timingLabel?: string,
|};

type GAPayload = GAEvent | GAPageView | GATiming;

export type GoogleAnalytics = ('send', GAPayload) => {};

export function sendAnalytics(payload: GAPayload) {
  const ga: ?GoogleAnalytics = self.ga;
  if (ga) {
    ga('send', payload);
  }
}
