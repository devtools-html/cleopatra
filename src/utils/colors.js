/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * These are the colors from Photon. They are inlined to provide easy access. If updating
 * please change the CSS variables as well.
 *
 * Firefox Colors v1.0.3
 * https://github.com/FirefoxUX/design-tokens/tree/master/colors
 */

export const MAGENTA_50 = '#ff1ad9';
export const MAGENTA_60 = '#ed00b5';
export const MAGENTA_70 = '#b5007f';
export const MAGENTA_80 = '#7d004f';
export const MAGENTA_90 = '#440027';

export const PURPLE_50 = '#9400ff';
export const PURPLE_60 = '#8000d7';
export const PURPLE_70 = '#6200a4';
export const PURPLE_80 = '#440071';
export const PURPLE_90 = '#25003e';

export const BLUE_40 = '#45a1ff';
export const BLUE_50 = '#0a84ff';
export const BLUE_60 = '#0060df';
export const BLUE_70 = '#003eaa';
export const BLUE_80 = '#002275';
export const BLUE_90 = '#000f40';

export const TEAL_50 = '#00feff';
export const TEAL_60 = '#00c8d7';
export const TEAL_70 = '#008ea4';
export const TEAL_80 = '#005a71';
export const TEAL_90 = '#002d3e';

export const GREEN_50 = '#30e60b';
export const GREEN_60 = '#12bc00';
export const GREEN_70 = '#058b00';
export const GREEN_80 = '#006504';
export const GREEN_90 = '#003706';

export const YELLOW_50 = '#ffe900';
export const YELLOW_60 = '#d7b600';
export const YELLOW_70 = '#a47f00';
export const YELLOW_80 = '#715100';
export const YELLOW_90 = '#3e2800';

export const RED_50 = '#ff0039';
export const RED_60 = '#d70022';
export const RED_70 = '#a4000f';
export const RED_80 = '#5a0002';
export const RED_90 = '#3e0200';

export const ORANGE_50 = '#ff9400';
export const ORANGE_60 = '#d76e00';
export const ORANGE_70 = '#a44900';
export const ORANGE_80 = '#712b00';
export const ORANGE_90 = '#3e1300';

export const GREY_10 = '#f9f9fa';
export const GREY_20 = '#ededf0';
export const GREY_30 = '#d7d7db';
export const GREY_40 = '#b1b1b3';
export const GREY_50 = '#737373';
export const GREY_60 = '#4a4a4f';
export const GREY_70 = '#38383d';
export const GREY_80 = '#2a2a2e';
export const GREY_90 = '#0c0c0d';

export const INK_70 = '#363959';
export const INK_80 = '#202340';
export const INK_90 = '#0f1126';
