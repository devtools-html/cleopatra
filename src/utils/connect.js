/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import type { Dispatch, State } from '../types/store';

type MapStateToProps<OwnProps: Object, StateProps: Object> = (
  state: State,
  ownProps: OwnProps
) => StateProps;

type MapDispatchToProps<OwnProps: Object, DispatchProps: Object> =
  | ((dispatch: Dispatch, ownProps: OwnProps) => DispatchProps)
  | DispatchProps;

type MergeProps<
  StateProps,
  DispatchProps: Object,
  OwnProps: Object,
  Props: Object
> = (
  stateProps: StateProps,
  dispatchProps: DispatchProps,
  ownProps: OwnProps
) => Props;

type ConnectOptions = {
  pure?: boolean,
  areStatesEqual?: boolean,
  areOwnPropsEqual?: boolean,
  areStatePropsEqual?: boolean,
  areMergedPropsEqual?: boolean,
  storeKey?: boolean,
  withRef?: boolean,
};

export type ExplicitConnectOptions<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> = {
  mapStateToProps?: MapStateToProps<OwnProps, StateProps>,
  mapDispatchToProps?: MapDispatchToProps<OwnProps, DispatchProps>,
  mergeProps?: MergeProps<
    StateProps,
    DispatchProps,
    OwnProps,
    ConnectedProps<OwnProps, StateProps, DispatchProps>
  >,
  options?: ConnectOptions,
  component: React.ComponentType<
    ConnectedProps<OwnProps, StateProps, DispatchProps>
  >,
};

export type ConnectedProps<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> = $ReadOnly<{|
  ...OwnProps,
  ...StateProps,
  ...DispatchProps,
|}>;

export type ConnectedComponent<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> =
  | React.ComponentType<ConnectedProps<OwnProps, StateProps, DispatchProps>>
  | React.StatelessFunctionalComponent<
      ConnectedProps<OwnProps, StateProps, DispatchProps>
    >;

/**
 * react-redux's connect function is too polymorphic and problematic. This function
 * is a wrapper to simplify the typing of connect and make it more explicit, and
 * less magical.
 */
export default function explicitConnect<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
>(
  connectOptions: ExplicitConnectOptions<OwnProps, StateProps, DispatchProps>
): React.ComponentType<OwnProps> {
  const {
    mapStateToProps,
    mapDispatchToProps,
    mergeProps,
    options,
    component,
  } = connectOptions;

  // Opt out of the flow-typed definition of react-redux's connect, and use our own.
  return (connect: any)(
    mapStateToProps,
    mapDispatchToProps,
    mergeProps,
    options
  )(component);
}
