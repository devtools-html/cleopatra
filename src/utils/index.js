/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals(a: Object, b: Object): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (a.hasOwnProperty(key)) {
      aLength++;
      if (a[key] !== b[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (b.hasOwnProperty(key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}
