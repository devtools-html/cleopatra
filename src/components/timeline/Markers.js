/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import classNames from 'classnames';
import { timeCode } from '../../utils/time-code';
import { withSize } from '../shared/WithSize';
import Tooltip from '../tooltip/Tooltip';
import { TooltipMarker } from '../tooltip/Marker';
import { markerStyles, overlayFills } from '../../profile-logic/marker-styles';
import explicitConnect from '../../utils/connect';
import { getPreviewSelection } from '../../selectors/profile';
import { getThreadSelectors } from '../../selectors/per-thread';
import { getSelectedThreadIndex } from '../../selectors/url-state';
import './Markers.css';

import type { Milliseconds, CssPixels } from '../../types/units';
import type { Marker, MarkerIndex } from '../../types/profile-derived';
import type { SizeProps } from '../shared/WithSize';
import type { ConnectedProps } from '../../utils/connect';
import type { ThreadIndex } from '../../types/profile';

// Exported for tests.
export const MIN_MARKER_WIDTH = 0.3;

type MarkerState = 'PRESSED' | 'HOVERED' | 'NONE';

/**
 * The TimelineMarkers component is built up of several nested components,
 * and they are all collected in this file. In pseudo-code, they take
 * the following forms:
 *
 * export const TimelineMarkersJank = (
 *  <Connect markers={JankMarkers}>
 *    <WithSize>
 *      <TimelineMarkers />
 *    </WithSize>
 *  </Connect>
 * );
 *
 * export const TimelineMarkersOverview = (
 *   <Connect markers={AllMarkers}>
 *     <WithSize>
 *       <TimelineMarkers />
 *     </WithSize>
 *   </Connect>
 * );
 */

export type OwnProps = {|
  +rangeStart: Milliseconds,
  +rangeEnd: Milliseconds,
  +threadIndex: ThreadIndex,
  +onSelect: any,
|};

export type StateProps = {|
  +additionalClassName?: ?string,
  +getMarker: MarkerIndex => Marker,
  +markerIndexes: MarkerIndex[],
  +isSelected: boolean,
  +isModifyingSelection: boolean,
  +testId: string,
|};

type Props = {|
  ...ConnectedProps<OwnProps, StateProps, {||}>,
  ...SizeProps,
|};

type State = {
  hoveredItem: Marker | null,
  mouseDownItem: Marker | null,
  mouseX: CssPixels,
  mouseY: CssPixels,
};

class TimelineMarkersImplementation extends React.PureComponent<Props, State> {
  _canvas: HTMLCanvasElement | null = null;
  _requestedAnimationFrame: boolean = false;
  state = {
    hoveredItem: null,
    mouseDownItem: null,
    mouseX: 0,
    mouseY: 0,
  };

  _takeCanvasRef = (c: HTMLCanvasElement | null) => {
    this._canvas = c;
  };

  _scheduleDraw() {
    window.requestAnimationFrame(() => {
      const c = this._canvas;
      if (c) {
        timeCode('TimelineMarkersImplementation render', () => {
          this.drawCanvas(c);
        });
      }
    });
  }

  _hitTest(e): Marker | null {
    const c = this._canvas;
    if (c === null) {
      return null;
    }

    const r = c.getBoundingClientRect();
    const {
      width,
      rangeStart,
      rangeEnd,
      getMarker,
      markerIndexes,
    } = this.props;
    const x = e.pageX - r.left;
    const y = e.pageY - r.top;
    const rangeLength = rangeEnd - rangeStart;
    const time = rangeStart + (x / width) * rangeLength;
    const onePixelTime = (rangeLength / width) * window.devicePixelRatio;

    // Markers are drawn in array order; the one drawn last is on top. So if
    // there are multiple markers under the mouse, we want to find the one
    // with the highest array index. So we walk the list of markers
    // from high index to low index, which is front to back in z-order.
    for (let i = markerIndexes.length - 1; i >= 0; i--) {
      const markerIndex = markerIndexes[i];
      const marker = getMarker(markerIndex);
      const { start, dur, name } = marker;
      const duration = Math.max(dur, onePixelTime);
      if (time < start || time >= start + duration) {
        continue;
      }
      const markerStyle =
        name in markerStyles ? markerStyles[name] : markerStyles.default;
      if (y >= markerStyle.top && y < markerStyle.top + markerStyle.height) {
        return marker;
      }
    }
    return null;
  }

  _onMouseMove = (event: SyntheticMouseEvent<>) => {
    const hoveredItem = this._hitTest(event);
    if (hoveredItem !== null) {
      this.setState({
        hoveredItem,
        mouseX: event.pageX,
        mouseY: event.pageY,
      });
    } else if (this.state.hoveredItem !== null) {
      this.setState({
        hoveredItem: null,
      });
    }
  };

  _onMouseDown = e => {
    const mouseDownItem = this._hitTest(e);
    this.setState({ mouseDownItem });
    if (mouseDownItem !== null) {
      if (e.target.setCapture) {
        e.target.setCapture();
      }
      e.stopPropagation();
    }
  };

  _onMouseUp = e => {
    const { mouseDownItem } = this.state;
    if (mouseDownItem !== null) {
      const mouseUpItem = this._hitTest(e);
      if (
        mouseDownItem === mouseUpItem &&
        mouseUpItem !==
          null /* extra null check because flow doesn't realize it's unnecessary */
      ) {
        const { onSelect, threadIndex } = this.props;
        onSelect(
          threadIndex,
          mouseUpItem.start,
          mouseUpItem.start + mouseUpItem.dur
        );
      }
      this.setState({
        hoveredItem: mouseUpItem,
        mouseDownItem: null,
      });
    }
  };

  _onMouseOut = () => {
    this.setState({
      hoveredItem: null,
    });
  };

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      prevProps !== this.props ||
      prevState.hoveredItem !== this.state.hoveredItem ||
      prevState.mouseDownItem !== this.state.mouseDownItem
    ) {
      this._scheduleDraw();
    }
  }

  render() {
    const {
      additionalClassName,
      isSelected,
      isModifyingSelection,
      threadIndex,
      testId,
    } = this.props;

    const { mouseDownItem, hoveredItem, mouseX, mouseY } = this.state;
    const shouldShowTooltip = !isModifyingSelection && !mouseDownItem;

    return (
      <div
        data-testid={testId}
        className={classNames(
          'timelineMarkers',
          additionalClassName,
          isSelected ? 'selected' : null
        )}
      >
        <canvas
          className="timelineMarkersCanvas"
          ref={this._takeCanvasRef}
          onMouseDown={this._onMouseDown}
          onMouseMove={this._onMouseMove}
          onMouseUp={this._onMouseUp}
          onMouseOut={this._onMouseOut}
        />
        {shouldShowTooltip && hoveredItem ? (
          <Tooltip mouseX={mouseX} mouseY={mouseY}>
            <TooltipMarker marker={hoveredItem} threadIndex={threadIndex} />
          </Tooltip>
        ) : null}
      </div>
    );
  }

  _drawRoundedRect(
    ctx: CanvasRenderingContext2D,
    x: CssPixels,
    y: CssPixels,
    width: CssPixels,
    height: CssPixels,
    cornerSize: CssPixels
  ) {
    // Cut out c x c -sized squares in the corners.
    const c = Math.min(width / 2, Math.min(height / 2, cornerSize));
    const bottom = y + height;
    ctx.fillRect(x + c, y, width - 2 * c, c);
    ctx.fillRect(x, y + c, width, height - 2 * c);
    ctx.fillRect(x + c, bottom - c, width - 2 * c, c);
  }

  drawCanvas(c: HTMLCanvasElement) {
    const {
      rangeStart,
      rangeEnd,
      width,
      getMarker,
      markerIndexes,
    } = this.props;

    const devicePixelRatio = c.ownerDocument
      ? c.ownerDocument.defaultView.devicePixelRatio
      : 1;
    const height = c.getBoundingClientRect().height;
    const pixelWidth = Math.round(width * devicePixelRatio);
    const pixelHeight = Math.round(height * devicePixelRatio);

    if (c.width !== pixelWidth || c.height !== pixelHeight) {
      c.width = pixelWidth;
      c.height = pixelHeight;
    }
    const ctx = c.getContext('2d');
    if (ctx === null || ctx === undefined) {
      return;
    }

    ctx.clearRect(0, 0, pixelWidth, pixelHeight);
    ctx.scale(devicePixelRatio, devicePixelRatio);

    let previousPos = null;
    for (const markerIndex of markerIndexes) {
      const marker = getMarker(markerIndex);
      const { start, dur, name } = marker;
      let pos = ((start - rangeStart) / (rangeEnd - rangeStart)) * width;
      pos = Math.round(pos * devicePixelRatio) / devicePixelRatio;

      if (previousPos === pos && dur === 0) {
        // This position has already been drawn, let's move to the next marker!
        continue;
      }
      previousPos = pos;
      const itemWidth = Number.isFinite(dur)
        ? Math.max(
            (dur / (rangeEnd - rangeStart)) * width,
            MIN_MARKER_WIDTH / devicePixelRatio
          )
        : Number.MAX_SAFE_INTEGER;
      const markerStyle =
        name in markerStyles ? markerStyles[name] : markerStyles.default;
      ctx.fillStyle = markerStyle.background;
      if (markerStyle.squareCorners) {
        ctx.fillRect(pos, markerStyle.top, itemWidth, markerStyle.height);
      } else {
        this._drawRoundedRect(
          ctx,
          pos,
          markerStyle.top,
          itemWidth,
          markerStyle.height,
          1 / devicePixelRatio
        );
      }
      if (markerStyle.borderLeft !== null) {
        ctx.fillStyle = markerStyle.borderLeft;
        ctx.fillRect(pos, markerStyle.top, 1, markerStyle.height);
      }
      if (markerStyle.borderRight !== null) {
        ctx.fillStyle = markerStyle.borderRight;
        ctx.fillRect(
          pos + itemWidth - 1,
          markerStyle.top,
          1,
          markerStyle.height
        );
      }
      const markerState = this._getMarkerState(marker);
      if (markerState === 'HOVERED' || markerState === 'PRESSED') {
        ctx.fillStyle = overlayFills[markerState];
        if (markerStyle.squareCorners) {
          ctx.fillRect(pos, markerStyle.top, itemWidth, markerStyle.height);
        } else {
          this._drawRoundedRect(
            ctx,
            pos,
            markerStyle.top,
            itemWidth,
            markerStyle.height,
            1 / devicePixelRatio
          );
        }
      }
    }
    ctx.scale(1 / devicePixelRatio, 1 / devicePixelRatio);
  }

  _getMarkerState(marker: Marker): MarkerState {
    const { hoveredItem, mouseDownItem } = this.state;
    if (mouseDownItem !== null) {
      if (marker === mouseDownItem && marker === hoveredItem) {
        return 'PRESSED';
      }
      return 'NONE';
    }
    if (marker === hoveredItem) {
      return 'HOVERED';
    }
    return 'NONE';
  }
}

/**
 * Combine the base implementation of the TimelineMarkers with the
 * WithSize component.
 */
export const TimelineMarkers = withSize<Props>(TimelineMarkersImplementation);

/**
 * Create a special connected component for Jank instances.
 */
export const TimelineMarkersJank = explicitConnect<OwnProps, StateProps, {||}>({
  mapStateToProps: (state, props) => {
    const { threadIndex } = props;
    const selectors = getThreadSelectors(threadIndex);
    const selectedThread = getSelectedThreadIndex(state);

    return {
      getMarker: selectors.getMarkerGetter(state),
      markerIndexes: selectors.getJankMarkerIndexesForHeader(state),
      isSelected: threadIndex === selectedThread,
      isModifyingSelection: getPreviewSelection(state).isModifying,
      testId: 'TimelineMarkersJank',
    };
  },
  component: TimelineMarkers,
});

/**
 * Create a connected component for all markers.
 */
export const TimelineMarkersOverview = explicitConnect<
  OwnProps,
  StateProps,
  {||}
>({
  mapStateToProps: (state, props) => {
    const { threadIndex } = props;
    const selectors = getThreadSelectors(threadIndex);
    const selectedThread = getSelectedThreadIndex(state);
    const markerIndexes = selectors.getCommittedRangeAndTabFilteredMarkerIndexesForHeader(
      state
    );
    return {
      additionalClassName:
        selectors.getThread(state).name === 'GeckoMain'
          ? 'timelineMarkersGeckoMain'
          : null,
      getMarker: selectors.getMarkerGetter(state),
      markerIndexes,
      isSelected: threadIndex === selectedThread,
      isModifyingSelection: getPreviewSelection(state).isModifying,
      testId: 'TimelineMarkersOverview',
    };
  },
  component: TimelineMarkers,
});

/**
 * FileIO is an optional marker type. Only add these markers if they exist.
 */
export const TimelineMarkersFileIo = explicitConnect<
  OwnProps,
  StateProps,
  {||}
>({
  mapStateToProps: (state, props) => {
    const { threadIndex } = props;
    const selectors = getThreadSelectors(threadIndex);
    const selectedThread = getSelectedThreadIndex(state);

    return {
      getMarker: selectors.getMarkerGetter(state),
      markerIndexes: selectors.getFileIoMarkerIndexes(state),
      isSelected: threadIndex === selectedThread,
      isModifyingSelection: getPreviewSelection(state).isModifying,
      testId: 'TimelineMarkersFileIo',
    };
  },
  component: TimelineMarkers,
});

/**
 * Create a component for memory-related markers.
 */
export const TimelineMarkersMemory = explicitConnect<
  OwnProps,
  StateProps,
  {||}
>({
  mapStateToProps: (state, props) => {
    const { threadIndex } = props;
    const selectors = getThreadSelectors(threadIndex);
    const selectedThread = getSelectedThreadIndex(state);

    return {
      getMarker: selectors.getMarkerGetter(state),
      markerIndexes: selectors.getMemoryMarkerIndexes(state),
      isSelected: threadIndex === selectedThread,
      isModifyingSelection: getPreviewSelection(state).isModifying,
      additionalClassName: 'timelineMarkersMemory',
      testId: 'TimelineMarkersMemory',
    };
  },
  component: TimelineMarkers,
});

/**
 * Create a component for IPC-related markers.
 */
export const TimelineMarkersIPC = explicitConnect<OwnProps, StateProps, {||}>({
  mapStateToProps: (state, props) => {
    const { threadIndex } = props;
    const selectors = getThreadSelectors(threadIndex);
    const selectedThread = getSelectedThreadIndex(state);

    return {
      getMarker: selectors.getMarkerGetter(state),
      markerIndexes: selectors.getIPCMarkerIndexes(state),
      isSelected: threadIndex === selectedThread,
      isModifyingSelection: getPreviewSelection(state).isModifying,
      additionalClassName: 'timelineMarkersIPC',
      testId: 'TimelineMarkersIPC',
    };
  },
  component: TimelineMarkers,
});
