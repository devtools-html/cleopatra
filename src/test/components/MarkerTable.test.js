/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import Markers from '../../components/marker-table';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/processed-profile';
import { getBoundingBox } from '../fixtures/utils';

describe('MarkerTable', function() {
  it('renders some basic markers', () => {
    // Set an arbitrary size that will not kick in any virtualization behavior.
    jest
      .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
      .mockImplementation(() => getBoundingBox(2000, 1000));

    // These were all taken from real-world values.
    const profile = getProfileWithMarkers(
      [
        [
          'UserTiming',
          12.5,
          {
            type: 'UserTiming',
            startTime: 12.5,
            endTime: 12.5,
            name: 'foobar',
            entryType: 'mark',
          },
        ],
        [
          'NotifyDidPaint',
          14.5,
          {
            type: 'tracing',
            category: 'Paint',
            interval: 'start',
          },
        ],
        [
          'setTimeout',
          165.87091900000001,
          {
            type: 'Text',
            startTime: 165.87091900000001,
            endTime: 165.871503,
            name: '5.5',
          },
        ],
      ]
        // Sort the markers.
        .sort((a, b) => a[1] - b[1])
    );

    const { container } = render(
      <Provider store={storeWithProfile(profile)}>
        <Markers />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
