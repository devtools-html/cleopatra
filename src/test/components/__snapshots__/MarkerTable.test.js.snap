// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerTable renders some basic markers 1`] = `
<div
  aria-labelledby="marker-table-tab-button"
  class="markerTable"
  id="marker-table-tab"
  role="tabpanel"
>
  <div
    class="markerSettings"
  >
    <div
      class="markerSettingsSpacer"
    />
    <div
      class="markerSettingsSearchbar"
    >
      <label
        class="markerSettingsSearchbarLabel"
      >
        Filter Markers: 
        <form
          class="idleSearchField markerSettingsSearchField"
        >
          <input
            class="idleSearchFieldInput photon-input"
            name="search"
            placeholder="Enter filter terms"
            required=""
            title="Only display markers that match a certain name"
            type="search"
            value=""
          />
          <input
            class="idleSearchFieldButton"
            tabindex="-1"
            type="reset"
          />
        </form>
      </label>
    </div>
  </div>
  <div
    class="treeView"
  >
    <div
      class="treeViewHeader"
    >
      <span
        class="treeViewHeaderColumn treeViewFixedColumn start"
      >
        Start
      </span>
      <span
        class="treeViewHeaderColumn treeViewFixedColumn duration"
      >
        Duration
      </span>
      <span
        class="treeViewHeaderColumn treeViewFixedColumn category"
      >
        Category
      </span>
      <span
        class="treeViewHeaderColumn treeViewMainColumn name"
      >
        Description
      </span>
    </div>
    <div
      class="react-contextmenu-wrapper treeViewContextMenu"
    >
      <div
        aria-label="Call tree"
        class="treeViewBody"
        role="tree"
        tabindex="0"
      >
        <div
          class="treeViewBodyInnerWrapper"
        >
          <div
            class="treeViewBodyInner treeViewBodyInner0"
            style="height: 48px;"
          >
            <div
              class="treeViewBodyInner treeViewBodyInner0TopSpacer"
              style="height: 0px;"
            />
            <div
              class="treeViewBodyInner treeViewBodyInner0InnerChunk"
            >
              <div
                class="treeViewRow treeViewRowFixedColumns even "
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeViewRowColumn treeViewFixedColumn start"
                  title="0.000s"
                >
                  0.000s
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn duration"
                  title="—"
                >
                  —
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn category"
                  title="UserTiming"
                >
                  UserTiming
                </span>
              </div>
              <div
                class="treeViewRow treeViewRowFixedColumns odd "
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeViewRowColumn treeViewFixedColumn start"
                  title="0.002s"
                >
                  0.002s
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn duration"
                  title="unknown"
                >
                  unknown
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn category"
                  title="Paint"
                >
                  Paint
                </span>
              </div>
              <div
                class="treeViewRow treeViewRowFixedColumns even "
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeViewRowColumn treeViewFixedColumn start"
                  title="0.153s"
                >
                  0.153s
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn duration"
                  title="0.001ms"
                >
                  0.001ms
                </span>
                <span
                  class="treeViewRowColumn treeViewFixedColumn category"
                  title="Text"
                >
                  Text
                </span>
              </div>
            </div>
          </div>
          <div
            class="treeViewBodyInner treeViewBodyInner1"
            style="height: 48px; width: 3000px;"
          >
            <div
              class="treeViewBodyInner treeViewBodyInner1TopSpacer"
              style="height: 0px;"
            />
            <div
              class="treeViewBodyInner treeViewBodyInner1InnerChunk"
            >
              <div
                aria-label="foobar, running time is undefined (undefined), self time is undefined"
                aria-level="1"
                aria-selected="false"
                class="treeViewRow treeViewRowScrolledColumns even  "
                id="treeViewRow-0"
                role="treeitem"
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeRowIndentSpacer"
                  style="width: 0px;"
                />
                <span
                  class="treeRowToggleButton collapsed leaf"
                />
                <span
                  class="treeViewRowColumn treeViewMainColumn name"
                >
                  foobar
                </span>
              </div>
              <div
                aria-label="NotifyDidPaint, running time is undefined (undefined), self time is undefined"
                aria-level="1"
                aria-selected="false"
                class="treeViewRow treeViewRowScrolledColumns odd  "
                id="treeViewRow-1"
                role="treeitem"
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeRowIndentSpacer"
                  style="width: 0px;"
                />
                <span
                  class="treeRowToggleButton collapsed leaf"
                />
                <span
                  class="treeViewRowColumn treeViewMainColumn name"
                >
                  NotifyDidPaint
                </span>
              </div>
              <div
                aria-label="setTimeout — 5.5, running time is undefined (undefined), self time is undefined"
                aria-level="1"
                aria-selected="false"
                class="treeViewRow treeViewRowScrolledColumns even  "
                id="treeViewRow-2"
                role="treeitem"
                style="height: 16px; line-height: 16px;"
              >
                <span
                  class="treeRowIndentSpacer"
                  style="width: 0px;"
                />
                <span
                  class="treeRowToggleButton collapsed leaf"
                />
                <span
                  class="treeViewRowColumn treeViewMainColumn name"
                >
                  setTimeout — 5.5
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
